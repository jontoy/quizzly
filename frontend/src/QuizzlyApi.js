import axios from "axios";
import config from "./config";
const DATABASE_URL =
  process.env.REACT_APP_BASE_URL ||
  config.DATABASE_URL ||
  "http://localhost:3001";

class QuizzlyApi {
  static async request(endpoint, paramsOrData = {}, verb = "get") {
    console.debug("API Call:", endpoint, paramsOrData, verb);

    try {
      return (
        await axios({
          method: verb,
          url: `${DATABASE_URL}/${endpoint}`,
          [verb === "get" ? "params" : "data"]: paramsOrData,
        })
      ).data;
      // axios sends query string data via the "params" key,
      // and request body data via the "data" key,
      // so the key we need depends on the HTTP verb
    } catch (err) {
      console.error("API Error:", err.response);
      let message = err.response.data.message;
      throw Array.isArray(message) ? message : [message];
    }
  }
  static async getQuizzes({ name, min_difficulty, max_difficulty }) {
    let res = await this.request(`quiz`, {
      name,
      min_difficulty,
      max_difficulty,
    });
    return res.quizzes;
  }
  static async getQuiz(id) {
    let res = await this.request(`quiz/${id}`);
    return res.quiz;
  }
  static async getQuizAnswers(id) {
    let res = await this.request(`quiz/${id}/answers`);
    return res.answers;
  }
  static async getQuestions({ text, quiz_id }) {
    let res = await this.request(`question`, {
      text,
      quiz_id,
    });
    return res.quizzes;
  }
  static async getQuestionsByQuizId(quiz_id) {
    return await this.getQuestions({ quiz_id });
  }
  static async getQuestion(question_id) {
    let res = await this.request(`question/${question_id}`);
    return res.question;
  }
  static async getOptions({ value, question_id, is_correct }) {
    let res = await this.request(`option`, {
      value,
      question_id,
      is_correct,
    });
    return res.quizzes;
  }
  static async getOptionsByQuestionId(question_id) {
    return await this.getOptions({ question_id });
  }
  static async getOption(option_id) {
    let res = await this.request(`option/${option_id}`);
    return res.option;
  }
}

export default QuizzlyApi;
