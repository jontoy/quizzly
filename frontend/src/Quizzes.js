import React, { useState, useEffect } from "react";
import QuizzlyApi from "./QuizzlyApi";
import Searchbox from "./Searchbox";
import QuizList from "./QuizList";

const Quizzes = () => {
  const [isLoading, setIsLoading] = useState(true);
  const [quizzes, setQuizzes] = useState([]);
  const [searchTerm, setSearchTerm] = useState("");
  useEffect(() => {
    async function getQuizzes() {
      let allQuizzes = await QuizzlyApi.getQuizzes({ name: searchTerm });
      setQuizzes(allQuizzes);
    }
    getQuizzes();
    setIsLoading(false);
  }, [searchTerm]);
  if (isLoading) {
    return <p>Loading &hellip;</p>;
  }

  return (
    <div className="Quizzes container">
      <Searchbox filterResults={setSearchTerm} />
      <QuizList quizzes={quizzes} />
    </div>
  );
};

export default Quizzes;
