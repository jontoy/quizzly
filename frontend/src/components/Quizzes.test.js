import React from "react";
import { render } from "@testing-library/react";
import Quizzes from "./Quizzes";
import { MemoryRouter } from "react-router-dom";
import QuizzlyApi from "../QuizzlyApi";

jest.mock("../QuizzlyApi");
QuizzlyApi.getQuizzes.mockResolvedValue([]);

it("renders without crashing", () => {
  render(
    <MemoryRouter>
      <Quizzes />
    </MemoryRouter>
  );
});

it("matches snapshot", () => {
  const { asFragment } = render(
    <MemoryRouter>
      <Quizzes />
    </MemoryRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});
