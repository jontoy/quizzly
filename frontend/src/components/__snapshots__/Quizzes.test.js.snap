// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Quizzes container"
  >
    <form
      class="Searchbox form-inline mt-3"
    >
      <input
        class="form-control form-control-lg flex-grow-1"
        name="search"
        placeholder="Search..."
        type="search"
        value=""
      />
    </form>
    <ul
      class="QuizList list-unstyled"
    />
  </div>
</DocumentFragment>
`;
